app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,31 @@
+class OmniauthCallbacksController < ApplicationController
+  def complete
+    auth = request.env['omniauth.auth']
+    email_address = auth.info.email
+    one_login_uid = auth.uid
+
+    candidate = Candidate.find_by(one_login_uid: one_login_uid) || Candidate.find_by(email_address: email_address)
+
+    if candidate
+      candidate.update!(one_login_uid: one_login_uid) if candidate.one_login_uid.nil?
+    else
+      candidate = Candidate.create(email_address: email_address, one_login_uid: one_login_uid)
+    end
+
+    session[:onelogin_id_token] = auth.credentials.id_token
+    candidate.update!(last_signed_in_at: Time.zone.now) if sign_in(candidate, scope: :candidate)
+
+    redirect_to root_path
+  end
+
+  def sign_out
+    id_token = session[:onelogin_id_token]
+    reset_session
```

Comment:
If we're going on the basis that this action reverts the above action, would it make sense to do the following?

```ruby
id_token = session.delete(:onelogin_id_token)
```

Reply:
Does DfE Sign In's implementation reset the entire session?